File: backend/poc_v3/src/embed_test.ts
import { matmul } from "@huggingface/transformers";
import {
  captionText,
  loadEmbedder,
  OPERATIONAL_SOURCES,
  type ItemInput,
} from "./embed.ts";

const MODEL_DIR = "./models/embeddinggemma-300m-onnx";
const e = await loadEmbedder(MODEL_DIR);

const items: ItemInput[] = [
  { text: "PaymentProcessor.charge() throws on null amount", source: "vscode" },
  { text: "fix the null-pointer in checkout", source: "user" },
  { text: "schedule lunch with Alex", source: "user" },
  { text: "Stripe 402 from billing endpoint", source: "terminal" },
  { text: "deploy the staging build to prod", source: "terminal" },
];

console.log("Captioning check:");
const captioned = items.map((it) => {
  const c = captionText(it);
  const isOp = it.source && OPERATIONAL_SOURCES.has(it.source);
  console.log(`  [${it.source ?? "-"}]${isOp ? " (op)" : ""}  ->  ${c}`);
  return c;
});

const emb = await e.encode(captioned);
console.log(`\nembed() returned tensor: dims=${emb.dims}, dtype=${emb.type}`);

// L2 normalization is baked into the community ONNX export, so for unit vectors
// dot product == cosine similarity, and matmul(emb, emb.T) gives the full
// similarity matrix in one call (mirrors the example in the model card).
const sims = (await matmul(emb, emb.transpose(1, 0))).tolist() as number[][];

const labels = items.map((_, i) => `q${i}`);
const short = (s: string) =>
  (s.length > 36 ? s.slice(0, 33) + "..." : s).padEnd(36);

console.log("\nCosine matrix (matmul, L2-normalized):");
process.stdout.write(" ".repeat(38));
for (const l of labels) process.stdout.write(l.padStart(8));
console.log();

items.forEach((item, i) => {
  process.stdout.write(short(`q${i}: ${item.text}`));
  for (let j = 0; j < items.length; j++) {
    process.stdout.write(sims[i]![j]!.toFixed(3).padStart(8));
  }
  console.log();
});

console.log("\nSpot checks (same as before):");
console.log("  q0 vs q1 (billing-related):  ", sims[0]![1]!.toFixed(3));
console.log("  q0 vs q2 (vs lunch):         ", sims[0]![2]!.toFixed(3));
console.log("  q0 vs q3 (vs Stripe 402):    ", sims[0]![3]!.toFixed(3));

console.log("\nRanking vs q0 (charge null):");
const ranked = sims[0]!
  .map((score, index) => ({ index, score }))
  .sort((a, b) => b.score - a.score);
for (const r of ranked) {
  console.log(`  q${r.index}: ${r.score.toFixed(3)}  ${items[r.index]!.text}`);
}

File: backend/poc_v3/scripts/embed_test.py
"""Python equivalent of src/embed_test.ts.

Loads the same EmbeddingGemma model via sentence-transformers (PyTorch) and
runs the same 5 captioned items through encode() to produce a similarity
matrix. Compare against the TS run to validate the model pipeline.

Note: this loads the ORIGINAL `google/embeddinggemma-300m` (PyTorch) rather
than the ONNX export. The weights are the same; only the runtime differs.
Expected: numbers should match the TS run to within float precision (~1e-4).
"""

from __future__ import annotations

import os
import numpy as np
from sentence_transformers import SentenceTransformer

os.environ.setdefault("HF_HUB_DISABLE_PROGRESS_BARS", "1")

OPERATIONAL_CAPTIONS = {
    "db": "SQL",
    "vscode": "Code",
    "terminal": "Command",
    "build": "Build",
}
OPERATIONAL_SOURCES = set(OPERATIONAL_CAPTIONS)


def caption_text(text: str, source: str | None) -> str:
    if not source:
        return text
    tag = OPERATIONAL_CAPTIONS.get(source)
    return f"{tag}: {text}" if tag else text


def main() -> int:
    if not os.environ.get("HF_TOKEN"):
        print("HF_TOKEN not set — model is gated, re-accept at", file=__import__("sys").stderr)
        print("https://huggingface.co/google/embeddinggemma-300m", file=__import__("sys").stderr)
        return 1

    print("Loading sentence-transformers model: google/embeddinggemma-300m")
    model = SentenceTransformer("google/embeddinggemma-300m")

    items = [
        ("vscode",   "PaymentProcessor.charge() throws on null amount"),
        ("user",     "fix the null-pointer in checkout"),
        ("user",     "schedule lunch with Alex"),
        ("terminal", "Stripe 402 from billing endpoint"),
        ("terminal", "deploy the staging build to prod"),
    ]

    print("\nCaptioning check:")
    captioned = []
    for source, text in items:
        c = caption_text(text, source)
        is_op = source in OPERATIONAL_SOURCES
        print(f"  [{source}]{' (op)' if is_op else ''}  ->  {c}")
        captioned.append(c)

    print("\nEncoding via model.encode(...)...")
    emb = model.encode(captioned, normalize_embeddings=True)
    print(f"embed returned ndarray: shape={emb.shape}, dtype={emb.dtype}")

    # L2-normalized embeddings: dot product == cosine similarity.
    sims = emb @ emb.T

    def short(s: str) -> str:
        return (s[:33] + "..." if len(s) > 36 else s).ljust(36)

    print("\nCosine matrix (matmul, L2-normalized):")
    header = " " * 38 + "".join(f"q{i}".rjust(8) for i in range(len(items)))
    print(header)
    for i, (src, txt) in enumerate(items):
        row = short(f"q{i}: {txt}")
        row += "".join(f"{sims[i, j]:.3f}".rjust(8) for j in range(len(items)))
        print(row)

    print("\nSpot checks (same as TS test):")
    print(f"  q0 vs q1 (billing-related):  {sims[0, 1]:.3f}")
    print(f"  q0 vs q2 (vs lunch):         {sims[0, 2]:.3f}")
    print(f"  q0 vs q3 (vs Stripe 402):    {sims[0, 3]:.3f}")

    print("\nRanking vs q0 (charge null):")
    ranking = sorted(enumerate(sims[0]), key=lambda x: -x[1])
    for idx, score in ranking:
        print(f"  q{idx}: {score:.3f}  {items[idx][1]}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
